apps/web/app/blog/posts/ai-powered-spreadsheet/page.mdx

```diff
@@ -0,0 +1,217 @@
+
+## Key Learnings
+
+- Markdown tables work way better than JSON as context
+- Nested arrays are hard for LLMs to patch correctly
+- AI can self-correct at runtime when tools return well structured errors
+- Mouse selection beats typing "A1:D5"
+
+Let me walk through each layer of the architecture.
+
+## Managing Tabs with Interactables
+
+<video
+  src="https://github.com/user-attachments/assets/39d1b5b0-cff2-4a57-b825-f6ae9785eec4"
+  controls
+/>
+
+Let me start with the simplest case: tab management.
+
+Tabs are just lightweight metadata: an ID and a name. The standard interactable pattern works great for this.
+
+The interactable wrapper basically subscribes to the tab store and publishes a simple state object whenever anything changes:
+
+```typescript
+const TabsInteractableWrapper = (props) => {
+  const publishState = useCallback(() => {
+    const store = useSpreadsheetTabsStore.getState();
+
+    const payload = {
+      tabs: store.tabs.map((tab) => ({
+        id: tab.id,
+        name: tab.name,
+      })),
+      activeTabId: store.activeTabId,
+    };
+
+    onPropsUpdate({ state: payload });
+  }, []);
+
+  // Subscribe to store changes
+  useEffect(() => {
+    const unsubscribe = useSpreadsheetTabsStore.subscribe(publishState);
+    return () => unsubscribe();
+  }, [publishState]);
+
+  return null; // This component doesn't render anything
+};
+
+export const InteractableTabs = withInteractable(TabsInteractableWrapper, {
+  componentName: "TabsState",
+  description: "Spreadsheet tab metadata (names and IDs only, no data)",
+  propsSchema: interactableTabsPropsSchema,
+});
+```
+
+This is way lighter than the full spreadsheet data, so it's fine to just emit it as structured metadata. The AI can see tab names, rename them, switch between them, whatever. And because it's using the standard interactable pattern, it all just works with Tambo's built-in machinery.
+
+[Full code: src/components/ui/interactable-tabs.tsx](https://github.com/michaelmagan/cheatsheet/blob/main/src/components/ui/interactable-tabs.tsx)
+
+## Passing Spreadsheet Data as Markdown
+
+I initially tried just passing the raw JSON of the spreadsheet data to the AI. It didn't work well. The model would get confused or hallucinate cell references.
+
+So I tried a few different formats and found that markdown tables worked surprisingly well. JSON is inherently noisy. All those brackets, quotes, and keys add a lot of extra tokens that don't really help the model understand the tabular structure. Markdown tables are much cleaner, and I think LLMs have probably seen millions of them in their training data, so they just naturally understand rows and columns.
+
+Tambo's context helper feature made this super easy. I just write a function that formats the data, register it, and it automatically gets included with every message. I don't have to think about when or how to inject that context. It's just always there.
```

Comment:
minor but tense is in consistent - "Tambo made this super easy..." (past tense) "I just write a function..."

"Tambo makes this super easy..."